src/history.rs

```
@@ -668,149 +751,89 @@ impl HistoryImpl {
             self.new_items.len() - self.first_unwritten_new_item_index
         );
 
-        // We want to rewrite the file, while holding the lock for as briefly as possible
-        // To do this, we speculatively write a file, and then lock and see if our original file changed
-        // Repeat until we succeed or give up
-        let Some(possibly_indirect_target_name) = history_filename(&self.name, L!("")) else {
+        // We want to rewrite the history file.
+        // To avoid issues with crashes during writing,
+        // we write to a temporary file and once we are done, this file is renamed such that it
+        // replaces the original history file.
+        // To avoid races, the original history file needs to be locked for the entire duration.
+        // Taking a shared lock first and later upgrading to an exclusive one could result in a
+        // deadlock, so we take an exclusive one immediately.
+        let Ok(Some(mut locked_history_file)) = LockedHistoryFile::new(
+            LockingMode::Exclusive,
+            &self.name,
+            OFlag::O_RDONLY | OFlag::O_CREAT,
+            HISTORY_FILE_MODE,
+        ) else {
+            // TODO: better error handling
```

what should we improve here?

---

Well, right now, when `save_internal_via_rewrite` returns, we have no idea if it succeeded. For this specific location, we don't know if the return value of `LockedHistoryFile::new` was an error or `Ok(None)`. The latter indicates that `history_filename` returned `None`. The return type for `history_filename` is problematic, because `None` can indicate that the file name passed to it (for some reason called `session_id` in the function) is empty, or that `path_get_data` failed.

An empty file name should never make it into `save_internal_via_rewrite` in the first place, so we might consider panicking in that situation. On the other hand, not being able to get the data directory for storing the history file is a runtime error which should be addressed (e.g. with an error message). Silently ignoring it is certainly not ideal.

---

I added more extensive error handling, which allows returning a meaningful error from `save_internal_via_rewrite` in most cases. This allows to log errors at its only call site, which also makes the behavior more consistent with `save_internal_via_appending`.